tower-grpc-interop/src/client.rs

```diff
@@ -216,6 +216,302 @@ fn assert_success(result: Result<Vec<TestAssertion>, Box<Error>>)
+                    RESPONSE_LENGTHS == actual_response_lengths.as_slice(),
+                    format!("{:?}={:?}", RESPONSE_LENGTHS, actual_response_lengths)
+                ),
+            ]
+        })
+        .then(&assert_success))
+}
+
+fn make_ping_pong_request(idx: usize) -> pb::StreamingOutputCallRequest {
+    let req_len = REQUEST_LENGTHS[idx];
+    let resp_len = RESPONSE_LENGTHS[idx];
+    pb::StreamingOutputCallRequest {
+        response_parameters: vec![
+            pb::ResponseParameters::with_size(resp_len)
+        ],
+        payload: Some(util::client_payload(req_len as usize)),
+        ..Default::default()
+    }
+}
+
+/// Contains the state that is threaded through the recursive "loop"
+/// of futures that performs the actual ping-pong with the server.
+struct PingPongState {
+    sender: futures::sync::mpsc::UnboundedSender<pb::StreamingOutputCallRequest>,
+    stream: Box<Stream<Item=pb::StreamingOutputCallResponse, Error=tower_grpc::Error>>,
+    responses: Vec<pb::StreamingOutputCallResponse>,
+    assertions: Vec<TestAssertion>,
+}
+
+type PingPongResponsesFuture = Box<Future<
+    Item=(Vec<pb::StreamingOutputCallResponse>, Vec<TestAssertion>),
+    Error=Box<Error>>>;
+/// Recursive function that takes the current PingPongState of the ping-pong
+/// operation and performs the rest of it. It returns a future of
+/// responses from the server and assertions that were performed in
+/// the process.
+fn perform_ping_pong(state: PingPongState) -> PingPongResponsesFuture {
+    let PingPongState { sender, stream, mut responses, mut assertions } = state;
+    Box::new(stream
+        .into_future()  // Take one element from the stream.
+        .map_err(|(err, _stream)| -> Box<Error> {
+            Box::new(err)
+        })
+        .and_then(|(resp, stream)| -> PingPongResponsesFuture {
+            if let Some(resp) = resp {
+                responses.push(resp);
+                if responses.len() == REQUEST_LENGTHS.len() {
+                    // Close the request stream. This tells the server that there
+                    // won't be any more requests.
+                    drop(sender);
+                    Box::new(stream
+                        .into_future()
+                        .map_err(|err| -> Box<Error> {
+                            Box::new(err.0)
+                        })
+                        .map(|(resp, _stream)| {
+                            assertions.push(test_assert!(
+                                "server should close stream after client closes it.",
+                                resp.is_none(),
+                                format!("resp={:?}", resp)
+                            ));
+                            (responses, assertions)
+                        }))
+                } else {
+                    sender.unbounded_send(
+                        make_ping_pong_request(responses.len())).unwrap();
```

Comment:
Nit: formatting here is kind of weird, I think it would look better as:
```rust
sender.unbounded_send(
    make_ping_pong_request(responses.len())
).unwrap();
```

Reply:
Heh, being used to the Google C++ code style it's your version that seems weird to me. But this isn't Google C++... Fixed.

Do you know if rustfmt is usable now? It would be nice to automatically resolve this kind of thing.